Refactor and test progressive probabilistic Hough

matthew-brett opened the pull request:
This is a refactoring and commenting of the Probabilistic Hough, along with suitable tests, as I get more sure of what we should be testing.

Meanwhile, it turns out that our algorithm is a more or less direct port of the OpenCV algorithm to Python — see <https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/hough.cpp#L490>; I've noted this.j

```release-note
Refactor probabilistic Hough for clarity and accuracy, and extend tests for
probabilistic Hough.  The output from the function will differ slightly from
prior releases, in that a) the `line_length` input argument previously
specified the minimum acceptable line length *in terms of the maximum of the
up-down or left-right extent* (Chebyshev distance); the refactor takes
`line_length` to be the minimum acceptable length in terms of Euclidean
distance, and b) there was a minor error in the previous implementation, such
that the Hough accumulator was incorrectly updated for points found to be in
a discovered line, but that had not previously been processed.  This will have
minor effects on the output.
```

matthew-brett committed:
Rename some variables.

matthew-brett committed:
Rename variables for clarity, add comments.

matthew-brett committed:
Refactor to more idiomatic use of Numpy arrays

matthew-brett committed:
Refactor and add basic tests.

matthew-brett committed:
Refactor and expand tests

matthew-brett committed:
Remove incorrect comment

In fact, we only use each pixel once, by exhausting the random
permutation of the valid pixels.

matthew-brett committed:
Add tests for some diagnonal lines.

matthew-brett committed:
Use and test diagonal line length

matthew-brett committed:
Shorter gap allowed for crossing lines.

Prevent spurious line fusion.

matthew-brett committed:
Test gap on single line.

matthew-brett committed:
Drop gap further for diagonal lines.

matthew-brett commented:
> I see this was pushed to a branch on `main`; no problem, but generally prefer keeping that tree clean.

You mean - a branch on the upstream repo - yes - sorry - I must have goofed using the `gh` cli.

matthew-brett commented:
> > We should either rewrite, or add direct attribution.
> 
> The license is compatible, so we should definitely add attribution.

Sorry to be ignorant, but, apart from at the top of the file - where does this attribution go?

> > I'm assuming the Euclidean length is the better metric - have implemented and tested that. Comments?
> 
> That makes more sense; does that correspond to the paper?

It's not stated explicitly, but implied by for example their test images, with many lines each of a standard length, at different angles.

> > I still don't have a clear idea as to why this is desirable. Will investigate further.
> 
> Last night, I was wondering how we calculate the step size needed to advance exactly one pixel when traversing lines. Curious whether this was related.

I'll study a little more and get back to you.

matthew-brett committed:
Store pixels on line, reset those

Avoid recalculating line.

matthew-brett committed:
Refactor line finding for clarity.

matthew-brett commented:
I've refactored the line tracking code to remove the `>> shift` fixed-floating-point stuff, because it was so confusing.   I think the code is a lot easier to read now.    Before this refactor, with test script below:

```
35 ms ± 263 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

With the refactor, it's slightly slower, but I think the price is worth paying for readability:

```
36.3 ms ± 259 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Timing script:

```python
import skimage as ski

img = ski.data.camera()
edges = ski.feature.canny(img, 2, 1, 25)

%timeit ski.transform.probabilistic_hough_line(edges, 10)
```

matthew-brett committed:
Add algorithm to defer points causing ties

Some points cause ties in the accumulator.

Sometimes this is because the points are on two possible lines, and
sometimes it is because, for the selected points, small variations in
angles catch some part of the line, and the accumulator picks this up as
nearby thetas.

Try deferring points causing threshold ties to the end of the point
list.

This problem was likely the cause of adaptation of the the simplest
initial test, by restricting thetas.  Remove this adaptation, it does
not appear to be necessary after the tie check.

matthew-brett committed:
Begin testing of known line images

matthew-brett committed:
Complete precision / recall tests.

matthew-brett committed:
Adapt threshold to length of lines.

matthew-brett committed:
Fix divide by zero warning for normalized vectors.

matthew-brett committed:
Add LICENSE attribution for probabilistic Hough

matthew-brett committed:
Allow ties

matthew-brett committed:
Try looking for ties and sequential thetas.

In practice, this makes very little difference as tested by the
`precision_recall` function (see
`tests/skimage/transforms/test_hough_transform.py`).

matthew-brett committed:
Remove tie mean fix

It made little differences, and adds to code complexity.

matthew-brett committed:
Defend line-search from random truncated lines

Use random seeding to restrict test cases to known good lines.

This is to avoid maybe 10% of cases (in our tests) where random searches
hit truncated line.

matthew-brett committed:
Coderabbit suggestions.

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into refactor-test-pp-hough

* origin/main: (27 commits)
  Move `_shared` to `_skimage2` (#8102)
  Port `pad_footprint` & `mirror_footprint` to `skimage2` (#8094)
  Add rescaling API to skimage2 (#8075)
  Move `skimage2` implementation into `_skimage2` (#8093)
  Undo (double) mirroring in `ski2.morphology.dilation` (#8060)
  Grammar (#8091)
  Move `ensure_spacing` into `skimage2` (#8067)
  Ensure that `skimage2` does not eagerly import `skimage` (#8087)
  Use lazy-loader 0.5 (#8080)
  MAINT: make ellipse fitting forward compatible  (#8054)
  Avoid deprecated assign to ndarray.shape (#8020)
  Avoid circular import in `feature/corner.py` (#8077)
  Turn off coderabbit auto-labeling & label checks (#8070)
  Fix GIL being re-enabled by C++ extensions (#8059)
  Fix conventions following docstub and misc. (#8055)
  Port grayscale morphology operators to skimage2 (#8046)
  Add revised `peak_local_max` to `skimage2` (#8039)
  Allow read-only arrays as input to remap (#7535)
  Add `prescale`  parameter to "blob functions" (#7858)
  Try to make coderabbit respect exisiting type labels (#8042)
  ...

matthew-brett committed:
Ruff changes

matthew-brett commented:
@greptileai

matthew-brett committed:
Update LICENSE.txt

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

matthew-brett committed:
Greptile-inspired test optimizations.

matthew-brett commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,207 @@ def test_probabilistic_hough_bad_input():
+    return out[:size]
+
+
+def _get_lims(shape, margins):
+    margins = np.array(margins)
+    if margins.shape == ():
+        margins = np.ones(len(shape)) * margins
+    return np.stack((margins, shape - margins), axis=1)
```
This seems confused - the only use of `_get_lims` is to produce start and end indices for valid lines, and theses are used in a comparison (>= st_lims etc).

matthew-brett commented:
No human review for 2.5 months, and review would at least benefit from good understanding of Hough algorithm, which would take some time and energy.   I propose we merge this in a week or so, unless we do get a human review, so I can keep working on Skimage2 / Hough stuff.

matthew-brett committed:
Update comments from discussion.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,207 @@ def test_probabilistic_hough_bad_input():
+    return _sort_lines(transform.probabilistic_hough_line(*args, **kwargs))
+
+
+def _check_seeded_lines(seed, size, expected, img, **kwargs):
```
I think this function makes the tests it's used in less readable. Together with the sort functions, there's three additional layers of abstraction. And the naming of these functions doesn't make it transparent. `_sorted_ph` isn't telling me, what it does.

`_check_seeded_lines` is only used in twice, so the DRY-effect is questionable. Would you be opposed to inline it? It's short and that way we uncouple tests that should be independent.

Otherwise, I'd recommend a small refactor: better names, docstrings and keyword-only parameters to make these a bit less opaque when reading through a unit test.

matthew-brett replied:
I had a look, and I tried refactoring, but it wasn't looking any better to me.   I've added some docstrings, and I've renamed `_sorted_ph` to `_sorted_phl` (Sorted Probabilistic Hough Line) - in the hope that it's clearer.

I do not want to inline `_check_seeded_lines` as it's involved enough that inlining would a) mean I'd have to repeat comments / docstrings etc and b) would implement the same logic in two places.   So I've just tried my best to be as explanatory as I could in the docstring.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,207 @@ def test_probabilistic_hough_bad_input():
+    # because of the random order of point selection that sometimes leads
+    # to identification of part-lines (where the identified line is close to,
+    # but not exactly matching the real line).
+    seed_gen = np.random.default_rng(seed)
```
Nitpick: but `rng = np.random.default_rng(seed)` seems to be a convention at this point.

matthew-brett replied:
Yes, except I wanted to make it clear that this was an RNG that would in turn generate seeds for another RNG.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -371,51 +372,66 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
-    cdef int shift = 16
-    cdef int good_line
+    cdef cnp.intp_t[:, ::1] line_ends = np.zeros((2, 2), dtype=np.intp)
+    cdef Py_ssize_t max_distance, rho_idx_offset, idx
+    cdef cnp.float64_t line_sin, line_cos, rho, slope
+    cdef Py_ssize_t j, x, y, x1, y1, px, py, rho_idx, max_theta_idx
+    cdef Py_ssize_t reverse, gap, x_len, y_len, n_pts
+    cdef cnp.int64_t value, max_value
+    cdef int x_delta_1, delta, offset, slope_delta
     cdef Py_ssize_t nlines = 0
-    cdef Py_ssize_t lines_max = 2 ** 15  # maximum line number cutoff
+    cdef Py_ssize_t lines_max = 2 ** 15  # maximum line number cutoff.
     cdef cnp.intp_t[:, :, ::1] lines = np.zeros((lines_max, 2, 2),
                                                 dtype=np.intp)
-    max_distance = 2 * <Py_ssize_t>ceil((sqrt(img.shape[0] * img.shape[0] +
-                                              img.shape[1] * img.shape[1])))
-    cdef cnp.int64_t[:, ::1] accum = np.zeros((max_distance, theta.shape[0]),
+    cdef cnp.intp_t[::1] rand_idxs
```
This block could use a small refactor:

- Separate statements that only declare a variable and its type from statements that allocate/assign to variables. A newline and sorting would address this.
- Split into smaller more digestible blocks.
- Spice up with a few concise comments describing the dimensions of allocated arrays.

Let me know if I should make a suggestion. But you still might be better able to write the comments.

matthew-brett replied:
I've tried to do that - let me know if you want more changes.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -371,51 +372,66 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
     cdef Py_ssize_t height = img.shape[0]
     cdef Py_ssize_t width = img.shape[1]
 
     # compute the bins and allocate the accumulator array
```
This comment is misleading. The accumulator is allocated way later I think?

matthew-brett replied:
Fixed.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -371,51 +372,66 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
     cdef Py_ssize_t width = img.shape[1]
 
     # compute the bins and allocate the accumulator array
     cdef cnp.ndarray[ndim=2, dtype=cnp.uint8_t] mask = \
```
```suggestion
    cdef cnp.uint8_t[:, ::-1] mask = \
```
I think memoryviews are preferred. Might even be faster.

matthew-brett replied:
I'll take your word for it.   Where does the preference for memoryviews come from?  (I've implemented the change in the meantime).

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -466,102 +481,74 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
-                x0 = (x0 << shift) + (1 << (shift - 1))
-
-            # pass 1: walk the line, merging lines less than specified gap
+            # Line equation is r = cos theta x + sin theta y.  Rearranging:
```
This is supposed to be
```suggestion
            # Line equation is r = cos(theta) x + sin(theta) y.  Rearranging:
```
right?

matthew-brett replied:
Yes, I've put the brackets in in my new changes.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -427,35 +443,34 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
-            # Note - this will become slow when we have removed many pixels
-            # from the mask.
+        for p_i in range(n_pts):
+            # Select random non-zero point (step 1 above).
```
Nitpick: Make these very useful references even  more prominent:

```suggestion
            # STEP 1: Select random non-zero point (see Notes).
```

lagru commented on src/skimage/transform/_hough_transform.pyx:
Weird that using `cnp.intp_t[::1]` doesn't seem to work. I get a compilation error: 
```plain
.../src/skimage/transform/_hough_transform.pyx:448:9: Invalid index for memoryview specified, type intp_t[::1]
```

matthew-brett replied:
Yes, I think it's this:  https://stackoverflow.com/a/53834638 . So, to do that indexing, it can't be a memoryview.

lagru replied:
Hmm, could be that there are some performance improvements to be had, but not critical to this PR.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -427,35 +443,34 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
             value = 0
-            max_value = threshold - 1  # Max value in accumulator.
-            max_theta_idx = -1
+            max_value = -1  # Max value in accumulator, start value.
```
Why can we remove `threshold` here?

matthew-brett replied:
Do you mean - why can we do `for` without the `count > 0` check?   Because we check further down whether we have any points remaining.  Are you thinking of the case where there are no valid points to start with?

lagru replied:
No, why did we need `+ threshold` before, why can we drop it now? 

Because I'm guessing that `threshold - 1` might hav been a performance optimization. This ensures that the assignments in the if-block https://github.com/scikit-image/scikit-image/blob/1a138a1d84f00becd1a9e4ae434be144164c4ea3/src/skimage/transform/_hough_transform.pyx#L508-L510

are only executed once, when the accumulator reaches the necessary threshold.

matthew-brett replied:
Oh - yes - maybe - but it's really a micro-optimization.   Do you think it's worth restoring it?   It's a little opaque.

lagru replied:
Yes, In this case I'd keep it and record our guess with a comment. Might explain the slightly poorer performance you mentioned?

matthew-brett replied:
I put the old value back with comments.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -427,35 +443,34 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
-                accum_idx = round((ctheta[j] * x + stheta[j] * y)) + offset
-                accum[accum_idx, j] += 1
-                value = accum[accum_idx, j]
+                rho = ctheta[j] * x + stheta[j] * y
```
It would be good to introduce the variable `rho` and its meaning further above. Later you are using `r` inside the _line equation_ instead. That's a bit confusing.

matthew-brett replied:
I have introduced it above in the new comments.

matthew-brett replied:
I have used `rho` in the equations below.

lagru commented on src/skimage/transform/_hough_transform.pyx:
Where is _step 3_?

matthew-brett replied:
It's implicit, because we loop through each pixel in the for loop, and we therefore cannot return to a previous pixel.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -466,102 +481,74 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
+            # If line_sin is 0, set to marker value of 99, otherwise value
+            # will be between -1 and 1.
+            slope = -line_cos / line_sin if line_sin else 99
+            x_delta_1 = fabs(slope) < 1  # Does x advance in steps of 1?
```
`x_delta_1` could use a clearer name to signal that this it is a bool. Maybe `dx_is_1`, `delta_x_is_1`, `x_step_is_1`?

matthew-brett replied:
Done.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -466,102 +481,74 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
+            # Pass 1: walk the line, merging lines less than specified gap
             # length (step 5 continued).
-            for k in range(2):
+            line_pixels[0, 0] = x  # Insert current point into pixel store.
+            line_pixels[0, 1] = y
+            line_ends[:, 0] = x
+            line_ends[:, 1] = y
```
These could really use comments describing the meaning of their dimensions.

matthew-brett replied:
I added some comments.  Do they help?

lagru replied:
Not sure. 😅 It's difficult to figure out what you changed since the last review.

For example consider this https://github.com/scikit-image/scikit-image/blob/1a138a1d84f00becd1a9e4ae434be144164c4ea3/src/skimage/transform/_hough_transform.pyx#L442-L443

which dimension differentiates x & y, which dimension start and stop?

lagru replied:
Might make this easier to parse if this was refactored into an inline-able function.

lagru replied:
But I'll leave it to you if you want to do that.

matthew-brett replied:
Given we're only doing it this once, I'd rather not move the logic elsewhere.

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into refactor-test-pp-hough

* origin/main:
  Port feature.canny to skimage2 (#8047)
  Revert "Add AI contribution policy (#7982)" (#8124)
  Add AI contribution policy (#7982)

matthew-brett committed:
Add docstrings to Hough tests for clarity

matthew-brett committed:
More comments for variables in Cython Hough file

matthew-brett committed:
Check all margins

matthew-brett committed:
Lars suggests using memoryviews

matthew-brett commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -466,102 +481,74 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
             # Confirm line length is sufficient.
-            good_line = (labs(line_end[3] - line_end[1]) >= line_length or
-                         labs(line_end[2] - line_end[0]) >= line_length)
-            if not good_line:
+            x_len = line_ends[1, 0] - line_ends[0, 0]  # pass 2 x - pass 1 x
+            y_len = line_ends[1, 1] - line_ends[0, 1]  # pass 2 y - pass 1 y
+            if sqrt(x_len * x_len + y_len * y_len) < line_length:
                 continue
```
@lagru @stefanv - I took this as a bug-fix, but I see that it probably is worth warning users.   What do you think?

matthew-brett replied:
I think this is the key question - whether we need to warn here, or provide a route to the old behavior (span rather than euclidean distance).   My general feeling is that it's not worth warning / providing old behavior - given the algorithm is random, I suspect users won't notice this change, and the previous behavior was unexpected - and arguably wrong - as it prefers lines along the axes to diagonal lines.   Then again, users will find they get slightly more diagonal lines than they were previously getting.   What do y'all think?

matthew-brett replied:
On reflection - there are now a few bug-fixes in the code - the one we've already merged, and another one here, where I've fixed the unvoting of previously voted pixels.   So the behavior is going to be different from previous behavior in any case.   I think we can just let the length change slide quietly under the radar, and label it a bug-fix.  I can believe anyone was relying on it ...  So I'm going to vote for no warning / flag for old behavior.

stefanv replied:
I think this is better than what we had.

matthew-brett committed:
Apply suggestions from code review

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

matthew-brett committed:
Further commenting and variable renames.

matthew-brett committed:
Fix use of infinity as marker

matthew-brett committed:
Revert attempt to use memoryview

matthew-brett committed:
Fix within logic (fix suggested by Greptile)

matthew-brett reviewed (commented):
Some comments on the comments :)

matthew-brett committed:
Refactor Hough somewhat, emphasize comments.

matthew-brett committed:
Add check for voted pixels; expand comments.

matthew-brett committed:
Satisfy check that iterations exhausted.

matthew-brett committed:
Use memoryview for mask; pass lines_max as param

matthew-brett committed:
Rename variables, add comments to hough_line

matthew-brett committed:
Use n_rhos variable in probabilistic hough

matthew-brett committed:
Now we never negative vote, use uint for accum

matthew-brett committed:
Run pre-commit

matthew-brett committed:
Clarify that line length is Euclidean

matthew-brett committed:
Rather trivial greptile review

matthew-brett commented:
I think this is one is ready to merge now.   @lagru - what do you think?

matthew-brett committed:
Linting fixes

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into refactor-test-pp-hough

* origin/main: (85 commits)
  Drop Python 3.11 Support (#8153)
  undo change to benchmark workflow
  address review
  Update pyodide CI test to match the pattern used by numpy (#8140)
  Update .github/workflows/_build_linux_for_python_x.yaml
  Update .github/workflows/_test_linux_for_python_x.yaml
  Suppress tifffile DeprecationWarning in test_plugin_deprecation_on_imread (#8138)
  fix yaml
  fix yaml
  Update .github/workflows/test-linux.yaml
  Update .github/workflows/_test_linux_for_python_x.yaml
  Update .github/workflows/test-linux.yaml
  Update .github/workflows/test-linux.yaml
  Move tests back to doc build
  Fill out .PHONY targets
  Apply suggestions from code review
  Remove unused imports (apply commit hooks)
  Improve inserted docstring note of `deprecated_func` (#8095)
  address review
  Update doc/source/index.rst
  ...

matthew-brett commented:
@lagru - any thoughts?  I think this one is ready.

lagru commented on src/skimage/transform/_hough_transform.pyx:
```diff
@@ -371,197 +382,217 @@ def _probabilistic_hough_line(cnp.ndarray img, Py_ssize_t threshold,
+    cdef cnp.int64_t value, max_value
+    cdef int dx_is_1, delta, offset, slope_delta
+    # Starting and ending x and y coordinates of current discovered line.
+    cdef cnp.intp_t[:, ::1] line_ends = np.zeros((2, 2), dtype=np.intp)
```
```suggestion
    # Starting and ending of current discovered line.
    # Shape: (start & stop, x & y), 
    cdef cnp.intp_t[:, ::1] line_ends = np.zeros((2, 2), dtype=np.intp)
```

lagru replied:
Actually, consider just using two variables here `line_start` & `line_stop`?

lagru replied:
Nitpick though!

matthew-brett committed:
Update src/skimage/transform/_hough_transform.pyx

Co-authored-by: Lars Grüter <lagru@mailbox.org>

matthew-brett commented:
Sorry - I couldn't find this in the Github interface - but looking at it -
separating the forward and reverse ends of the line would make it harder to
do the loop logic where we update the forward in one pass of the loop, and
the reverse in the second :
https://github.com/scikit-image/scikit-image/blob/1a138a1d84f00becd1a9e4ae434be144164c4ea3/src/skimage/transform/_hough_transform.pyx#L561

matthew-brett committed:
Rewrite explanation of line_ends

matthew-brett reviewed (commented):
Comment on comment ...

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into refactor-test-pp-hough

* origin/main:
  Use standard C99 static inline instead of `__inline` (#8159)
  Fix structural_similarity win_size / gaussian_weights conflict (#8162)
  Unwrap seed refactor (#8142)
  ci: Remove llvm build and add ccache support on windows (#8164)
  Document warn_external in contributing guide (#8163)
  CI: Fix emscripten dependencies and clean up wheel test commands (#8147)
  Test only changed subpackages (#8103)
  Fix nightlies: update deprecated scipy call, fix underdetermined test
  Fix minkowski distanced invocation for SciPy >= 1.18 (#8157)
  Port `structural_similarly` to ski2 (#8104)
  make spin build requirement a lower bound (#8158)

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into refactor-test-pp-hough

* origin/main:
  Pin minimum pytest to 9 (#8174)
  docs: make source install instructions more discoverable and add build requirements to top level requirements file (#8165)
  Drop `--maxfail=5` from local pytest config (#8170)

stefanv committed:
Merge branch 'main' into refactor-test-pp-hough

matthew-brett commented:
@lars - can we get this over the line in the next couple of days?

matthew-brett commented:
Plea for merge before Monday, so we can do the big skimage2 port.

lagru commented:
I'm on it and will review asap – tomorrow at the latest. But of course I can't commit to approving. 

I'm don't think this should hold up the port, and I wouldn't know why the refactor should be hard to reconcile with `main` after the port? Otherwise, I'm happy to figure it out.

lagru committed:
Use more expressive names in test helpers

lagru committed:
Group cases in `test_probabilistic_hough_examples` with newlines

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,262 @@ def test_probabilistic_hough_bad_input():
+    # points - line from [0, 0] through [0, 3] is length 3.
+    assert _sorted_probabilistic_hough_line(img, line_length=L) == []
+
+    # Single line in UD (y)
```
What does "UD" stand for? Can we use more expressive names? Same for "RL" (right–left?) above.

Maybe a reminder to be a bit more conservative about abbreviations, especially if they are used only once.

matthew-brett replied:
I've expanded the comments a bit.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,262 @@ def test_probabilistic_hough_bad_input():
+    )
+
+    # Filter by length of diagonals.  Get only the diagonals back.
+    len_diag = int(np.floor(np.sqrt(2 * (n - 1) ** 2)))
```
I don't follow how the length of the shortest diagonal (?) is derived here?

matthew-brett replied:
The lines are diagonal, with LR and UD length `n`.  I've added some more comments to explain.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,262 @@ def test_probabilistic_hough_bad_input():
+    #
+    # """
+    # A hundred images containing a fixed number of randomly positioned lines
+    # were syntheticaly generated, the only source of noise being in the
```
Is this a citation? Is the typo in "syntheticaly" derived from the source?

matthew-brett replied:
Yes, the `"""` was meant to signal a direct quotation.   In fact it's from a pre-print - I've corrected the source of the reference, and fixed the typo (yes in the original) so the reader doesn't get tripped up.

lagru commented on tests/skimage/transform/test_hough_transform.py:
This test file really is a maze of helper functions and huge incremental tests. 😅

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -49,23 +49,17 @@ def test_probabilistic_hough():
+    _check_seeded_probabilistic_hough_lines(
+        seed=1964,
+        n_batches=50,
+        expected=[[(25, 25), (74, 74)], [(25, 75), (74, 26)]],
+        img=img,
+        threshold=10,
+        line_gap=1,
+        line_length=10,
     )
-    # sort the lines according to the x-axis
-    sorted_lines = []
-    for ln in lines:
-        ln = list(ln)
-        ln.sort(key=lambda x: x[0])
-        sorted_lines.append(ln)
 
-    assert [(25, 75), (74, 26)] in sorted_lines
-    assert [(25, 25), (74, 74)] in sorted_lines
```
This code hasn't been replaced with an equivalent one. Why do you drop the parameter `theta` here?

matthew-brett replied:
Do you mean - why did I not do the restriction of theta done in the original code?   Because that seemed to me a hack to get the code to pass, and I wanted to get the code to run without that hack.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,262 @@ def test_probabilistic_hough_bad_input():
+    tp = np.sum(detected_tf & out_tf)
+    fp = np.sum(detected_tf & ~out_tf)
+    denom = tp + fp  # Can be 0
+    return tp / denom if denom else 0.0, tp / np.sum(out_img)
```
I assume `tp` and `fp` are to be true and false positives respectively?

lagru replied:
I would have been significantly faster and more confident in figuring out what you are doing here if this was written like this:

```suggestion
    precision = tp / denom if denom else 0.0
    recall = tp / np.sum(out_img)
    return precision, recall
```

I had to look at the [calling context of the function](ttps://github.com/scikit-image/scikit-image/blob/1a6352de605660d07126808fcf34544633ea729f/tests/skimage/transform/test_hough_transform.py#L231) to confirm what is calculated here.

It might be good to dial down on the terseness a bit? 😅 In instances like this I don't think it's just a style choice.

lagru commented on tests/skimage/transform/test_hough_transform.py:
```diff
@@ -92,6 +86,262 @@ def test_probabilistic_hough_bad_input():
+    assert recall >= 0.75
+
+
+def precision_recall(lines, shape, threshold, rng=None):
```
```suggestion
def precision_recall(lines, shape, threshold, rng=None):
    """Calculate precision and recall of `probabilistic_hough_line` for given lines."""
```

lagru reviewed (approved):
In a perfect world with more time, this could use even more refactoring for readability. But I think by this point it's clear that we have already spent too much time on this (in context of the skimage2 transition).

Especially the test suite seems harder than necessary to follow with its many helper functions. Some of it was already there, some of it was added here. 

But I think the changes to the Hough implementation itself are a huge improvement! And all of the changes seem correct. So I think it's ready for inclusion as is. Leaving merging to you @matthew-brett so you have time to chime in with still open questions.

Thanks for bearing with me.

matthew-brett committed:
Restore microoptimization for threshold test

matthew-brett committed:
Apply suggestions from code review

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

matthew-brett committed:
Extend comments.

matthew-brett reviewed (commented):
Comments on comments, and references to new changes.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

Read on for the rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
